### PR TITLE
[WIP] quantization with tokamax ragged_dot

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

Start with a short description of what the PR does and how this is a change from
the past.

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/123456
FIXES: #123456

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [ ] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [ ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have run end-to-end tests tests and provided workload links above if applicable.
- [ ] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/layers/moe.py [hunk 1]
```diff
@@ -809,50 +808,41 @@ def gmm(inputs, kernel, tiling, group_sizes, expert_assignments):
           min(tiling[1], k),
           min(tiling[2], n),
       )
-      if self.config.use_tokamax_gmm:
-        output = tokamax.ragged_dot(
```
Comment (BirdsOfAFthr) [inline src/MaxText/layers/moe.py:813]: So you mean bf16 will also use the customized ops.py?

I don't think we want that?

File: src/MaxText/layers/moe.py
Comment (BirdsOfAFthr) [inline src/MaxText/layers/moe.py:1]: We need to update sharding for FP8

File: src/MaxText/kernels/megablox/ops.py [hunk 2]
```diff
@@ -94,15 +103,17 @@ def _gmm_fwd(
...
   if quantization_rule:
     if quantization_rule.act_qtype:
-      lhs = qpl.quantize(
-          lhs,
-          quantization_rule.act_qtype,
-          channelwise_axes=[] if quantization_rule.disable_channelwise_axes else [0],
-          calibration_method=quantization_rule.act_calibration_method,
-          scale_dtype=jnp.float32,
-      )
+      with set_xla_metadata(MUST_FUSE=fwd_counter):
```
Comment (BirdsOfAFthr) [inline src/MaxText/kernels/megablox/ops.py]: remove `MUST_FUSE` related code. here and elsewhere

File: src/MaxText/kernels/megablox/ops.py [hunk 3]
```diff
@@ -114,18 +125,38 @@ def _gmm_fwd(
...
-      group_offset,
-      existing_out,
-      transpose_rhs=transpose_rhs,
-      interpret=interpret,
-  )
+      # QAG is only supported for following conditions
+      if quantization_rule.weight_calibration_method.startswith("fixed") and isinstance(rhs, qpl.QArray):
+        rhs_qvalue = jax.lax.all_gather(rhs.qvalue, "fsdp", axis=0, tiled=True)
+        rhs = dataclasses.replace(rhs, qvalue=rhs_qvalue)
+  if use_tokamax_backend:
```
Comment (BirdsOfAFthr) [inline src/MaxText/kernels/megablox/ops.py:124]: Are we removing this in a followup?

There are too many knobs to follow. We don't need the old backend.py. I think we can completely remove it

File: src/MaxText/kernels/megablox/backend.py
Comment (BirdsOfAFthr) [inline src/MaxText/kernels/megablox/backend.py:1]: No need to change this file.

File: src/MaxText/kernels/megablox/ops.py [hunk 4]
```diff
@@ -173,41 +207,76 @@ def _gmm_bwd(
...
-      lhs.swapaxes(0, 1),
-      drhs_dout,
-      group_sizes,
-      rhs_dtype,
-      tiling,
-      group_offset,
-      num_actual_groups,
-      interpret=interpret,
-  )
+  if quantization_rule and quantization_rule.bwd_qtype:
```
Comment (BirdsOfAFthr) [inline src/MaxText/kernels/megablox/ops.py]: The same condition for qag applies here
`if quantization_rule.weight_calibration_method.startswith("fixed") and isinstance(rhs, qpl.QArray):`

File: src/MaxText/layers/moe.py [hunk 5]
```diff
@@ -809,50 +808,41 @@ def gmm(inputs, kernel, tiling, group_sizes, expert_assignments):
...
+      if self.config.megablox:
```
Comment (RissyRan) [inline src/MaxText/layers/moe.py]: Oh, please be careful about this change. We previously update to `use_tokamax_gmm` for less confusion in configs.

Comment (khatwanimohit) [reply src/MaxText/layers/moe.py]: use_tokamax_gmm is passed to this codepath now and there are conditions in ops.py on whether to use our implementation of megablox or tokamax's ragged_dot_general. 

Is it still concerning ?

**Review by BirdsOfAFthr — COMMENTED** [review]
It seems that there are no inputs for the static calibration method

**Comment by khatwanimohit** [comment]
will open a separate PR
